### ISSUE TITLE
Bounty for the found issue

### ISSUE THREAD

**Issue by annadanchenko** [status-im/ideas#11]
## Preamble

    Idea: DEV#11
    Title: Bounty for the found issues
    Status: Draft
    Created: 2017-11-13

## Summary

Make it possible to give bounties for the found bugs

## Vision

Anyone who reported an issue can be rewarded with a bounty if the issue is valuable. 

1. Contributor creates new issue in the GitHub project with mandatory text `bug-bounty`.
2. Team member reviews bug and if it's a valuable one then assigns label `bug-bounty` to the issue. This triggers a bounty contract to be created on the issue. 
3. Team member adds a label to classify the importance of the bug (more important -> bigger bounty). E.g `XS`, `S`, `M` and `L`
4. Team treasurer sends tokens (e.g. SNT) to the contract.
5. Payment 
- option 1: Repository owner can confirm the payment as soon as contract has funds (got SNT)
- option 2: Payment can be triggered only after issue is fixed (PR is merged), to promote important bugs to be reported first.

We trust that team members will evaluate the bug and pay proper reward for it (rewards should be listed per project probably, e.g XS in project A is 2 000 SNT, XS in project B is 100 SNT, there is no voting mechanism. Contributors will just see the history for each bug in GH and decide if they want to participate in the project. So if reward is considered unfair then no more contributions will come.
 
## Swarm Participants

- Lead Contributor: @annadanchenko
- Testing & Evaluation: 
- Treasurer: @andytudhope
- Contributor: @asemiankevich 
- Contributor: @antdanchenko 
- Contributor: @lukaszfryc   

## Requirements

1. Contributors should be able to discover such bug-bounty projects. Status Open Bounty site needs to have separate sections for listing/searching the projects.
2. Organisations who need bug hunters should be able to add, suspend and remove "bug-bounty" for the repositories.
3. Extra information on the exact build number(s), urls, expectations should be provided by an Organisation to the contributors to target their efforts to the right version in the right time. Their should be a common shared "template" and "location" for each project where this info is visible.
4. Need to prevent bug reports submitted as "stubs" with no useful content that will be modified later and more valuable bug report content copied to such "stub" (so it looks like user A added same bug before user B). Possible solution: as soon as new is issue created with text `bug-bounty` bot would copy it's tittle and content as a comment, so reviewer can see initial bug report.
5. Issue can be fixed, verified and closed even before we send funds, so need to take into account closed issues 
 
### Minimum Viable Product

Goal Date: 2018-03-30
Description: 
In short: contributor can get test SNT for the reported bug. All process is manual (to verify it works and find weak spots before implementing smart contracts and changing SOB site)
1. Contributor provides his/her ETH address on SOB site 
2. Contributor reports an issue in the GitHub repo and specifies `Bug-Bounty` type in the issue description
3. Reviewer adds a comment to the issue and if bug report is ok then add a label `bug-bounty`  and label `size` (e.g. `bug-bounty` and `M`). If bug report not ok then explain why (ask for extra details, close if duplicate with previous issue number, etc)
4. Team treasurer reviews new issues with "bug-bounty" label and sends SNT to the contributor's ETH address. Also adds a comment with transaction details
5. All bug hunting is done on specific build(s) within specified timeframe and functional scope (to be documented by bug hunt lead)

There will be a "closed beta" group of testers to avoid any frauds and games with copying others bug reports. If it works well and we implement copying initial issue description as a comment then next iterations can be open for everyone.

MVP verifies that proposed process can work and is aimed to find weak spots in the process before implementing bot/smart contracts/SOB site modifications. 

Outcomes:
1. Process description for the next iteration
2. Description of the features to be implemented in future iteration(s)
3. Found Issues 

Swarm members:
Need several reviewers (2-3) who know open issues or can learn them to review new bugs for duplicates and good description. Also evaluator is needed to decide if swarm does what it promised.

### Iteration 1..N
Goal Date:  
Description: 
1. Add to the SOB site sections for Bug hunters: projects available, top hunters
2. Make a beta run to check this model on Status app or SOB 
3. Possibility to suspend the bug hunting - need more thinking. Same for removing project from the listing
4. Implement a bot who will copy the initial bug report description
5. Auto-funding? 

## Supporting Role Communication

 Finance: will use funds from SOB funding, if 24 000 SNT are used before end of iteration then we stop cc @andytudhope 

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by 3esmit** [comment]
I've been thinking about this idea since I've participated in ether-camp bug hunt bounty. 
I think would be nice if repo-owner be able to define which labels give funds for the issue reporter to withdraw.
For example, 
1. you define at SOB that in your repo, everytime someone label some issue with "security" the reward is 10000 SNT, "annoyance" 1000 SNT. 
2. someone post an issue
3. contributor of repo labels it as "security"
4. repo owner verifies and allows withdraw by the issue poster

**Comment by naghdy** [comment]
Interesting idea. From an abuse perspective, do you think this incentivizes people submit high volumes of low impact bugs, just so that they can make a bit of money?

**Comment by 3esmit** [comment]
@naghdy  I don't think this would be an issue, at least they are submiting bugs. I think that that if someone spots a bug it would be wanting to report it fast, faster then the others. So bug hunters would need to collaborate to attack the repository, I think it's an unlikely scenario, such as "developer gangs submitting only many low impact bugs and selling the high impact bugs to the black market". Would be a funny story BTW :laughing:

**Comment by naghdy** [comment]
Closing issue as this seems to not be active anymore.

**Comment by asemiankevich** [comment]
I am in for 10 h/w approximately for this idea

**Comment by antdanchenko** [comment]
10h per week starting from `when it's needed`

**Comment by lukaszfryc** [comment]
5 h/w if you need my help

**Comment by martinklepsch** [comment]
@annadanchenko While this is primarily intended for testers I also realized it might be useful for other contributors. Generally we want to reward people to write up well-defined bounties regardless of whether they are bugs, removal of technical debt or features. Probably this is a good first start and we could experiment with compensating additional issues later on.

**Comment by oskarth** [comment]
Closing this issue as part of spring cleaning. If this idea is still relevant, please submit a PR per https://github.com/status-im/ideas/#contributing. If you feel closing this issue is a mistake, feel free to re-open.
